api/services/user.service.js

```diff
@@ -36,45 +40,93 @@ export const register = async (req, res, next) => {
-  const isUserUsernameAlreadyExist = await userRepository.findUserByUsername(
-    username
-  );
-  const addressData = {
-    address,
-    city,
-    country,
-    state_province,
-    zipcode,
-  };
-
-  if (isUserEmailAlreadyExist) {
-    if (isUserUsernameAlreadyExist) {
-      return next(new Conflict("This username and email already exists!"));
+  var updatedUser = null;
+  try {
+    const user = await userRepository.findUserById(userId);
+    if (!user) {
+      return next(Conflict.invalidUser());
+    }
+    if (user.username !== username || user.email !== email) {
+      const isUserEmailAlreadyExist = await userRepository.findUserByEmail(
+        email
+      );
+      const isUserUsernameAlreadyExist =
+        await userRepository.findUserByUsername(username);
+      if (isUserEmailAlreadyExist) {
+        if (isUserUsernameAlreadyExist) {
+          return next(new Conflict("This username and email already exists!"));
+        }
+        return next(Conflict.emailAlreadyExists());
+      }
+      if (isUserUsernameAlreadyExist) {
+        return next(Conflict.usernameAlreadyExists());
+      }
+    }
+    var hashedPassword = password;
+    if (user.password !== password) {
+      hashedPassword = await hashPassword(password, 10);
     }
-    return next(Conflict.emailAlreadyExists());
-  }
-  if (isUserUsernameAlreadyExist) {
-    return next(Conflict.usernameAlreadyExists());
-  }
 
-  const hashedPassword = await hashPassword(password, 10);
-  const newAddress = await addressService.addAddress(addressData);
-  const userData = {
+    var addressId = await addressService.getAddressIdByName(address);
+    const addressData = {
+      address,
+      city,
+      country,
+      state_province,
+      zipcode,
+    };
+    if (!addressId) {
+      const newAddress = await addressService.addAddress(addressData);
+      addressId = newAddress.id;
+    } else {
+      const oldAddress = await addressService.getAddressById(addressId);
+      //here we prevent database query!
+      if (
+        oldAddress.address !== address ||
+        oldAddress.city !== city ||
+        oldAddress.country !== country ||
+        oldAddress.state_province !== state_province ||
+        oldAddress.zipcode !== zipcode
+      ) {
+        await addressService.updateAddress(addressData, addressId);
+      }
+    }
+    const userData = {
       firstName,
       lastName,
       middle: middleName,
       avatar,
-      addressID: newAddress.id,
+      addressID: addressId,
       phone,
       email,
       username,
       password: hashedPassword,
     };
-  await userRepository.addUser(userData);
+    //TODO update address information before updating the user!!!
+    updatedUser = await userRepository.updateUser(userData, userId);
+  } catch {
+    return next(new UnexpectedError());
+  }
 
-  return res.status(201)
+  res.status(200).json(updatedUser);
+};
+export const deleteUser = async (req, res, next) => {
```

Comment:
do we need to be able to remove users?

Reply:
yes administrator should have this feature

Reply:
Also if user wants to delete his account permanently from our system